react/react.d.ts

```diff
@@ -41,14 +41,8 @@ declare module React {
         propTypes?: ValidationMap<P>;
     }
 
-    interface ComponentClassType<P> extends ComponentStatics<P> {
-        new (): ComponentClass<P>;
-    }
-
-    interface ComponentClass<P> {
-        // Deprecated in 0.12. See http://fb.me/react-legacyfactory
-        // new(props: P): ReactElement<P>;
-        // (props: P): ReactElement<P>;
+    interface ComponentClass<P> extends ComponentStatics<P> {
+        new (): Component<P>;
```

Comment:
I was going to argue with you about new() returning a ReactElement, but I just tested and right now you're correct. However, in this case I would argue that we should model not what React actually does, but what the intention is (and what React will eventually do). I think I'm going to have a hard time explaining/making my case but I'll do my best. I know you know most/all of this stuff but I think it will help me explain my position.

```
// v0.11
interface ComponentClass<P> extends ComponentStatics<P> {
  new (props: P): ReactElement<P>; // this worked but was not intended for use
  (props: P): ReactElement; // this is what you were supposed to do (factory function)
}
```

So in v0.11, you were not supposed to do `new MyComponent({})` but rather just `MyComponent({})`, even if the former worked.

```
// v0.12
interface ComponentClass<P> extends ComponentStatics<P> {
  new (props: P): ReactElement<P>; // this worked but was not intended for use
  (props: P): ReactElement; // this is now deprecated (but still works)
}
```

In v0.12 they started the transition to being able to use ES6 classes. `createElement` was introduced which takes, for all intents and purposes, a JS class and creates an element out of it. So even though the result of `createClass` is under the hood what is modeled above, the actual usage of it is not. The usage is as if the return value from `createClass` was, indeed, a class. In TypeScript, the way to model classes is with `new ()`. So by doing `new (): Component<P>` we can more accurately model the intended (and future) usage. The point is not so much to allow users to do `new MyComponent()` as it is to allow users to do `class MyComponent extends React.Component<P>` and have TypeScript understand that `MyComponent` is congruous with `React.ComponentClass<P>`. 

```
// v0.13
interface ComponentClass<P> extends ComponentStatics<P> {
  new (): Component<P>; // this now works like it should
  (): ??; // this throws an error 
}
```

Already with v0.13 `createClass` returns an actual JS class and usage as a factory function throws.

My point here is that `new (): Component<P>` is the correct signature even if under the hood it's not. The only reason it's not correct under the hood is because v0.12 is a transition version. But modeling it as `new (): Component<P>` breaks nothing in v0.12 and gets us ready for v0.13.

Regarding deprecation notices, I'm good with leaving the deprecation notice on `(props: P): ReactElement<P>`. I can also add a comment to `new (): Component<P>` mentioning that it is not meant to be called.